### DISCUSSION TITLE
Publishing LL-HLS DVR links with time stamps

### DISCUSSION THREAD

**Discussion by tonywingmah** [OvenMediaLabs/OvenMediaEngine#1225] [Ideas]
Can OME record HLS live for 1 week to disk and publish links that can access 2 hours from 9am to 11AM?

**Comment by getroot** [comment]
HLS Dump saves m3u8 files and segments to disk. 
https://airensoft.gitbook.io/ovenmediaengine/rest-api/v1/virtualhost/application/stream/hls-dump


You will need to use a file based web server like nginx to publish this. 
If you want to control access from 9:00 to 11:00, your WAS should provide such functionality.

**Reply by tonywingmah** [reply]
I have read the document on the dump to disk and I am a little confused. Does dump to disk mean save to FMP4 file?

Here is our sports workflow
1. We record Badminton courts with security cameras 24 x 7 for the entire tournament (live HLS DVR)
2. We publish segments 9am-11am, 12pm-2pm, 3pm-5pm on our website
3. When the users goes to our website, they can click on links for the 9am, 12pm or 3pm and play.
4. Coaches can download the matches they want.

The reason why we do this:
1. It requires very little staff and almost no training to operate
a.) There is no camera operator required
b.) No need to make any changes to the streaming Server as the feeds are all continuous 24x7 HLS live streams
c.) If the match starts late or is extra long all that is required is the link needs to be changed.
d.) Volunteers can download all the matches for vod using our website
e.) When the tournament is over the 24x7 HLS segments are deleted 
f.) If someone wants to watch after the tournament is over they can go to the client's website. OME would not be used. 

An example of how the web link might look like:
http://host:8080/hls/tennis/index.m3u8?startunix=1682422299&duration=3600
The HLS playback will start at the Unix Epoch time and play for 3600 seconds and stop.

Is this something that OME will consider implementing?

**Reply by getroot** [reply]
Thanks to your explanation, I clearly understand what your requirements are.
OME provides DVR function. We call this Live Rewind.

https://airensoft.gitbook.io/ovenmediaengine/streaming/low-latency-hls#live-rewind

However, there is no function to publish to VoD (with #EXT-X-ENDLIST tag) specifying the playback start position from the live URL. Since this is a feature we can implement and it looks useful, we'll add it to our backlog. (This is a low priority because it doesn't seem to be a feature that many people need.)

And the download function is not clear. To provide downloads, files split into multiple segments will have to be combined and provided, but I don't think it's OME's role. It would be better to provide this using an external tool such as FFMPEG.

**Reply by tonywingmah** [reply]
You understand clearly. We use ffmpeg for download and it is very good. We now send hls with start and stop tag to FFMPEG and it will download the file. We are willing to pay for development of link to publish live VOD. Can we discuss privately?

**Reply by getroot** [reply]
Please send your suggestions to support@airensoft.com. thank you

**Comment by getroot** [comment]
You can implement this yourself by making good use of the recording function currently provided by OME without waiting for new features to be implemented in Live Rewind.

https://airensoft.gitbook.io/ovenmediaengine/recording-experiment

Since the recording result is output to the information file, you can implement VoD service using this file.

https://airensoft.gitbook.io/ovenmediaengine/recording-experiment#appendix-a.-recorded-file-information-specification

The absolute time is output to this file as shown below.

```
<createdTime>2020-12-04T12:53:51.455+0900</createdTime>
<startTime>2020-12-04T12:53:51.612+0900</startTime>
<finishTime>2020-12-04T12:54:51.473+0900</finishTime>
```

**Reply by tonywingmah** [reply]
Can you clarify? Is this live VOD? We don't want files as sports never stop on time.

**Reply by getroot** [reply]
OME's recording function provides segmentation recording. You can instruct OME to continuously generate TS files every 5 minutes from a specific live channel. Users watch live channels via OME. And when the user wants to watch the broadcast from 1:00 to 3:00 yesterday, he clicks another link and plays the VoD through your WAS.Your WAS creates an m3u8 from OME's recorded info file and delivers it to the user.
Of course this should implement a lot in your WAS.
